applications/StructuralMechanicsApplication/custom_constitutive/linear_j2_plasticity_3d.cpp

```diff
@@ -0,0 +1,464 @@
+        double derivative_yieldfunction = -2. * mu * (1. + kp_new / (3. * mu));
+        dgamma = dgamma - yieldfunction / derivative_yieldfunction;
+        mAccumulatedPlasticStrain =
+            mAccumulatedPlasticStrainOld + 0.8164965809277260 * dgamma;
+        norm_yieldfunction = std::abs(yieldfunction);
+    }
+    // TODO (marcelo): handle the case when no convergence is achieved.
+    return dgamma;
+}
+
+double LinearJ2Plasticity3D::yieldFunction(const double norm_dev_stress,
+                                           const Properties& rMaterialProperties)
+{
+    double yield_stress = rMaterialProperties[YIELD_STRESS];
+    double hardening_modulus = rMaterialProperties[ISOTROPIC_HARDENING_MODULUS];
+    double theta = rMaterialProperties[REFERENCE_HARDENING_MODULUS];
+    double delta_k = rMaterialProperties[INFINITY_HARDENING_MODULUS];
+    double hardening_exponent = rMaterialProperties[HARDENING_EXPONENT];
+    double k_old =
+        yield_stress + (theta * hardening_modulus * mAccumulatedPlasticStrainOld) +
+        (delta_k) * (1. - std::exp(-hardening_exponent * mAccumulatedPlasticStrainOld));
+
+    return norm_dev_stress - k_old * 0.8164965809277260; // sqrt(2/3)
+}
+
+void LinearJ2Plasticity3D::CalculateElasticMatrix(Matrix &D, const Properties &props)
+{
+    double E = props[YOUNG_MODULUS];
+    double poisson_ratio = props[POISSON_RATIO];
+    double lambda =
+        E * poisson_ratio / ((1. + poisson_ratio) * (1. - 2. * poisson_ratio));
+    double mu = E / (2. + 2. * poisson_ratio);
+
+    D.clear();
+    D.resize(6, 6, false);
+    D = ZeroMatrix(6, 6);
+
+    D(0, 0) = lambda + 2. * mu;
+    D(0, 1) = lambda;
+    D(0, 2) = lambda;
+    D(1, 0) = lambda;
+    D(1, 1) = lambda + 2. * mu;
+    D(1, 2) = lambda;
+    D(2, 0) = lambda;
+    D(2, 1) = lambda;
+    D(2, 2) = lambda + 2. * mu;
+    D(3, 3) = mu;
+    D(4, 4) = mu;
+    D(5, 5) = mu;
+}
+
+void LinearJ2Plasticity3D::CalculateTangentTensor(
+    double dgamma, double norm_s_trial, const Vector& N_new, const Properties& props, Matrix& D)
+{
+    double hardening_modulus = props[ISOTROPIC_HARDENING_MODULUS];
+    double theta = props[REFERENCE_HARDENING_MODULUS];
+    double delta_k = props[INFINITY_HARDENING_MODULUS];
+    double hardening_exponent = props[HARDENING_EXPONENT];
+    double E = props[YOUNG_MODULUS];
+    double poisson_ratio = props[POISSON_RATIO];
+    double mu = E / (2. + 2. * poisson_ratio);
+    double volumetric_modulus = E / (3. * (1. - 2. * poisson_ratio));
+
+    double kp_new = (theta * hardening_modulus) +
+                    delta_k * (hardening_exponent *
+                               std::exp(-hardening_exponent * mAccumulatedPlasticStrain));
+
+    double theta_new = 1 - (2. * mu * dgamma) / norm_s_trial;
+    double theta_new_b = 1. / (1. + kp_new / (3. * mu)) - (1. - theta_new);
+
+    D(0, 0) = volumetric_modulus + (2 * mu * theta_new * 2. / 3.) -
```

Comment:
I would group common operations

Reply:
Example: const double 2_mu_theta_new_b = 2 * mu * theta_new_b;

Reply:
I am not an expert, I don't know if the compiler recognizes common operations, do you know @RiccardoRossi ?